ocaml-lsp-server/src/ocamlformat_rpc.ml

```diff
@@ -0,0 +1,172 @@
+            [ ("module-item-spacing", "compact"); ("margin", "64") ]
+            client
+        with
+        | Error (`Msg msg) ->
+          Log.log ~section:"ocamlformat" (fun () ->
+              Log.msg "An error occured while configuring ocamlformat"
+                [ ("msg", `String msg) ])
+        | Ok () -> ())
+  with
+  | Error `Stopped -> Fiber.return ()
+  | Ok res -> (
+    let+ res = Scheduler.await_no_cancel res in
+    match res with
+    | Ok s -> s
+    | Error e -> Exn_with_backtrace.reraise e)
+
+let start_exn ?io_thread t =
+  let prog = "ocamlformat-rpc" in
+  match Bin.which prog with
+  | None ->
+    t := Binary_not_found;
+    Fiber.return @@ Error `No_process
+  | Some bin_path -> (
+    let bin = Fpath.to_string bin_path in
+    let argv = [| bin |] in
+    let pid, stdout, stdin =
+      let stdin_i, stdin_o = Unix.pipe () in
+      let stdout_i, stdout_o = Unix.pipe () in
+      let pid = Unix.create_process bin argv stdin_i stdout_o Unix.stderr in
+      (pid, stdout_i, stdin_o)
+    in
+    let open Fiber.O in
+    let input = Unix.in_channel_of_descr stdout in
+    let output = Unix.out_channel_of_descr stdin in
+    let* io_thread =
+      (* Don't create a new thread if there is already one. *)
+      (* This can happen if the Ocamlformat server crashed. *)
+      match io_thread with
+      | Some io_thread -> Fiber.return io_thread
+      | None -> Scheduler.create_thread ()
+    in
+    let* client = pick_client ~pid input output io_thread in
+    match client with
+    | Error (`Msg msg) ->
+      (* The process did start but something went wrong when negociating the
+         version so we need to kill it *)
+      Unix.kill pid Sys.sigkill;
+      Scheduler.stop io_thread;
+      Fiber.return @@ Error (`Msg msg)
+    | Ok client ->
+      let process = { pid; input; output; io_thread; client } in
+      let+ () = configure process in
+      Log.log ~section:"ocamlformat" (fun () ->
+          Log.msg "Ocamlformat-RPC server started" [ ("pid", `Int pid) ]);
+      t := Running process;
+      Ok process)
+
+let start ?io_thread t =
+  let open Fiber.O in
+  try
+    let+ result = start_exn ?io_thread t in
+    match result with
+    | Error (`Msg msg) ->
+      Log.log ~section:"ocamlformat" (fun () ->
+          Log.msg "An error occured while initializing ocamlformat"
+            [ ("msg", `String msg) ]);
+      Error `No_process
+    | r -> r
+  with
+  | Unix.Unix_error (e, _, _) ->
+    t := Errored;
+    Log.log ~section:"ocamlformat" (fun () ->
+        Log.msg "An error occured when starting ocamlformat"
+          [ ("msg", `String (Unix.error_message e)) ]);
+    Fiber.return @@ Error `No_process
+
+let get_process t =
+  let open Fiber.O in
+  match !t with
+  | Stopped -> start t
+  | Running ({ pid; io_thread; _ } as p) -> (
+    (* Is the process still running ? *)
+    match
+      Scheduler.async io_thread (fun () -> Unix.waitpid [ WNOHANG ] pid |> fst)
```

Comment:
You should use `Scheduler.wait_for_process` here. We do the waiting for all processes in a special thread.

Reply:
I tried, but it raise the following exception if I use it twice:
```
[Trace - 10:37:25 AM] Received response 'textDocument/hover - (10)' in 2ms. Request failed: uncaught exception (-32603).
Error data: {
    "exn": "File \"fiber-unix/src/scheduler.ml\", line 418, characters 28-34: Assertion failed",
    "backtrace": 
"Raised at Fiber_unix__Scheduler.Process_watcher.Process_table.add in file \"fiber-unix/src/scheduler.ml\", line 418, characters 28-40
Called from Fiber_unix__Scheduler.Process_watcher.register in file \"fiber-unix/src/scheduler.ml\", line 441, characters 4-31
Called from Fiber_unix__Scheduler.wait_for_process in file \"fiber-unix/src/scheduler.ml\", line 558, characters 2-71
Called from Fiber.Execution_context.run_jobs in file \"submodules/dune/src/fiber/fiber.ml\", line 204, characters 8-13\n"
}```

It looks like `Swheduler.waitpid` always tries to add the process to the watchlist `Table` and that fails when it is already in the table.

Reply:
Hum I think I see it: it also uses `Unix.waitpid` which is blocking. So in fact I should not try to use it like the non-blocking `Unix.waitpid [ WNOHANG ]`.

I should not call `Scheduler.wait_for_process` each time I want to query ocf to check that it is still running.
I should use `Scheduler.wait_for_process` as a watcher that will do something if ocf stops running.

Is that right ?

Reply:
dune.ml should give you an example of how it should work. Basically, when your RPC is in the running state, it should be roughly this:

```ocaml
Fiber.fork_and_join_unit
  (fun () -> (* code to interact with ocamlformat rpc *))
  (fun () -> Scheduler.wait_for_process pid)
```

Additionally, there should be a `Ocamlformat_rpc.run` function that will return a fiber that should complete when RPC is completely finished and everything is cleaned up.

Let me know if that's not clear and we can walk through this in a call.